fw/http_frame.c

```diff
@@ -2095,7 +2104,31 @@ do {									\
 			return r;
 		}
 
-		fallthrough;
+		T_FSM_JMP(HTTP2_SEND_FRAMES);
+	}
+
+	T_FSM_STATE(HTTP2_MAKE_TRAILER_FRAMES) {
+		CALC_SND_WND_AND_SET_FRAME_TYPE(HTTP2_HEADERS);
+		r = tfw_h2_insert_frame_header(sk, ctx, stream, frame_type,
+					       snd_wnd, stream->xmit.t_len);
+		if (unlikely(r)) {
+			T_WARN("Failed to make trail headers frame %d", r);
+			return r;
+		}
+
+		T_FSM_JMP(HTTP2_SEND_FRAMES);
+	}
+
+	T_FSM_STATE(HTTP2_MAKE_TRAILER_CONTINUATION_FRAMES) {
```

Comment:
I add test for long trailers https://github.com/tempesta-tech/tempesta-test/pull/651/commits/ec625cefffced6fd6e0334d5cebf08c545e59ba3 and it doesn't pass

Reply:
It's not due to these statements, but `ss_skb_cutoff_data`.

```
[10845.573649] Call Trace:
[10845.575340]  <IRQ>
[10845.577160]  dump_stack+0x96/0xc4
[10845.578844]  skb_fragment.cold+0x5/0x31 [tempesta_fw]
[10845.580546]  ss_skb_cutoff_data+0x696/0x6e0 [tempesta_fw]
[10845.582240]  ? ss_skb_list_chop_head_tail+0x1d0/0x1d0 [tempesta_fw]
[10845.584021]  ? __asan_storeN+0x12/0x20
[10845.585705]  ? eb_delete+0x203/0x210 [tempesta_lib]
[10845.587407]  tfw_http_msg_cutoff_body_chunks+0x6d/0xc0 [tempesta_fw]
[10845.589112]  tfw_h2_stream_xmit_prepare_resp+0x3d8/0x410 [tempesta_fw]
[10845.590800]  tfw_h2_make_frames+0x7dc/0x9d0 [tempesta_fw]
[10845.592491]  ? tcp_established_options+0x1ab/0x330
[10845.594184]  ? tfw_h2_frame_process+0xc80/0xc80 [tempesta_fw]
[10845.595847]  ? bpf_skops_hdr_opt_len+0x210/0x210
[10845.597516]  tfw_sk_fill_write_queue+0x323/0x3e0 [tempesta_fw]
[10845.599158]  ? tfw_cli_conn_release+0xc0/0xc0 [tempesta_fw]
[10845.600788]  ? tfw_h2_sched_activate_stream+0xc8/0x190 [tempesta_fw]
[10845.602427]  ? ss_skb_tcp_entail_list+0x25/0x1a0 [tempesta_fw]
[10845.604115]  ss_tx_action+0xacf/0xc90 [tempesta_fw]
[10845.605703]  ? ss_skb_tcp_entail_list+0x1a0/0x1a0 [tempesta_fw]
[10845.607384]  ? __kernel_fpu_begin_mask+0x9a/0x150
[10845.609071]  net_tx_action+0xfa/0x3c0
[10845.610741]  __do_softirq+0xfe/0x383
[10845.612321]  asm_call_irq_on_stack+0x12/0x20
[10845.613910]  </IRQ>

```

```c
static int
skb_fragment(struct sk_buff *skb_head, struct sk_buff *skb, char *pspt,
             int len, TfwStr *it, int *fragn)
{
        if (unlikely(abs(len) > PAGE_SIZE)) {
                T_WARN("Attempt to add or delete too much data: %u\n", len);
                return -EINVAL;
        }
...
```

Reply:
No matter whether one trailer header is too long or too much trailer headers, as long as the size of the body plus trailers exceeds a `PAGE_SIZE`, it will fail at `tfw_http_msg_cutoff_body_chunks`, so this is not the bug of this PR, instead, we need to fix it individually.

I hacked the trailer `HEADER_FRAME` size to 100 bytes so the other trailer data could get packed into `CONTINUATION_FRAME`. It passes the test. So the trailer continuation framing logic has no problem.

Please resolve this conversation.

```diff
diff --git a/fw/http_frame.c b/fw/http_frame.c
index 94ccee5e..b6a71d5f 100644
--- a/fw/http_frame.c
+++ b/fw/http_frame.c
@@ -2114,7 +2114,7 @@ do {                                                                      \
        T_FSM_STATE(HTTP2_MAKE_TRAILER_FRAMES) {
                CALC_SND_WND_AND_SET_FRAME_TYPE(HTTP2_HEADERS);
                r = tfw_h2_insert_frame_header(sk, ctx, stream, frame_type,
-                                              snd_wnd, stream->xmit.t_len);
+                                              snd_wnd, 100);
                if (unlikely(r)) {
                        T_WARN("Failed to make trail headers frame %d", r);
                        return r;

```

```diff
diff --git a/encoding/test_encoding.py b/encoding/test_encoding.py
index 5bf2662..bda4e0b 100644
--- a/encoding/test_encoding.py
+++ b/encoding/test_encoding.py
@@ -526,8 +526,12 @@ class TestH2ChunkedWithTrailer(tester.TempestaTest, CommonUtils):

     def setUp(self):
         self.backends = copy.deepcopy(self.backends_template)
+        trailers = f"X-Token: {self.token}\r\n"
+        for i in range(100):
+            trailers += f"foobar-{i}: {'a'*10}\r\n"
+        trailers += "\r\n"
         self.backends[0]["response_content"] += (
-            self.encode_chunked(BODY_PAYLOAD, CHUNK_SIZE)[:-2] + f"X-Token: {self.token}\r\n\r\n"
+            self.encode_chunked(BODY_PAYLOAD, CHUNK_SIZE)[:-2] + f"{trailers}"
         )
         super().setUp()

```

```bash
$ t encoding.test_encoding.TestH2ChunkedWithTrailer
Root privileges are required: need access to module loading on localhost.

----------------------------------------------------------------------
Running functional tests...
----------------------------------------------------------------------

test (encoding.test_encoding.TestH2ChunkedWithTrailer) ... ok

----------------------------------------------------------------------
Ran 1 test in 2.308s

OK

```

```
[73739.183457] [tempesta fw] Tempesta FW is ready
[73739.235065] ---> HTTP2_MAKE_TRAILER_CONTINUATION_FRAMES
```